http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyClientTlsAuthTest.java

```diff
@@ -0,0 +1,176 @@
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
+import software.amazon.awssdk.http.async.SdkAsyncHttpResponseHandler;
+import software.amazon.awssdk.utils.AttributeMap;
+
+/**
+ * Tests to ensure that Netty layer can perform TLS client authentication.
+ */
+public class NettyClientTlsAuthTest extends ClientTlsAuthTestBase {
+    private static final AttributeMap DEFAULTS = AttributeMap.builder()
+            .put(TRUST_ALL_CERTIFICATES, true)
+            .build();
+
+    private static WireMockServer mockProxy;
+    private static ProxyConfiguration proxyCfg;
+    private static TlsKeyManagersProvider keyManagersProvider;
+
+    private SdkAsyncHttpClient netty;
+
+    @BeforeClass
+    public static void setUp() throws IOException {
+        ClientTlsAuthTestBase.setUp();
+
+        // Will be used by both client and server to trust the self-signed
+        // cert they present to each other
+        System.setProperty("javax.net.ssl.trustStore", serverKeyStore.toAbsolutePath().toString());
+        System.setProperty("javax.net.ssl.trustStorePassword", STORE_PASSWORD);
+        System.setProperty("javax.net.ssl.trustStoreType", "jks");
+
+        mockProxy = new WireMockServer(new WireMockConfiguration()
+                .dynamicHttpsPort()
+                .needClientAuth(true)
+                .keystorePath(serverKeyStore.toAbsolutePath().toString())
+                .keystorePassword(STORE_PASSWORD));
+
+        mockProxy.start();
+
+        mockProxy.stubFor(get(urlPathMatching(".*")).willReturn(aResponse().withStatus(200).withBody("hello")));
+
+        proxyCfg = ProxyConfiguration.builder()
+                .scheme("https")
+                .host("localhost")
+                .port(mockProxy.httpsPort())
+                .build();
+
+        keyManagersProvider = FileStoreTlsKeyManagersProvider.create(clientKeyStore, CLIENT_STORE_TYPE, STORE_PASSWORD);
+    }
+
+    @AfterClass
+    public static void teardown() throws IOException {
+        ClientTlsAuthTestBase.teardown();
+
+        mockProxy.stop();
+
+        System.clearProperty("javax.net.ssl.trustStore");
+        System.clearProperty("javax.net.ssl.trustStorePassword");
+        System.clearProperty("javax.net.ssl.trustStoreType");
+    }
+
+    @After
+    public void methodTeardown() {
+        if (netty != null) {
+            netty.close();
+        }
+        netty = null;
+    }
+
+    @Test(expected = IOException.class)
+    public void proxyRequest_ableToAuthenticate() throws Throwable {
+        netty = NettyNioAsyncHttpClient.builder()
+                .proxyConfiguration(proxyCfg)
+                .tlsKeyManagersProvider(keyManagersProvider)
+                .buildWithDefaults(DEFAULTS);
+
+        try {
+            sendRequest(netty, new RecordingResponseHandler());
+        } catch (CompletionException e) {
+            Throwable cause = e.getCause();
```

Comment:
This rule applies to all your tests that expect exceptions. Consider using junit 4 @Rule ExpectedException instead of this manual try catch testing stuff.

Reply:
Cool! Done.